src/junit.rs

```diff
@@ -7,42 +7,81 @@ use quick_xml::events::{BytesStart, Event};
 use quick_xml::reader::Reader;
 
 use crate::compute_name::{compute_name, unescape_str};
-use crate::testrun::{check_testsuites_name, Framework, Outcome, ParsingInfo, Testrun};
+use crate::testrun::{check_testsuites_name, Framework, Outcome, Testrun};
 use crate::validated_string::ValidatedString;
-
-#[derive(Default)]
-struct RelevantAttrs {
-    classname: Option<ValidatedString>,
-    name: Option<ValidatedString>,
-    time: Option<String>,
-    file: Option<ValidatedString>,
-}
+use crate::warning::WarningInfo;
 
 fn convert_attribute(attribute: Attribute) -> Result<String> {
     let bytes = attribute.value.into_owned();
     let value = String::from_utf8(bytes).context("Error converting attribute to string")?;
     Ok(value)
 }
+struct TestcaseAttrs {
+    name: ValidatedString,
+    time: Option<String>,
+    classname: Option<ValidatedString>,
+    file: Option<ValidatedString>,
+}
 
-fn validate_attribute(attribute: Attribute) -> Result<ValidatedString> {
-    ValidatedString::from_string(convert_attribute(attribute)?)
-        .context("Error validating attribute")
+enum AttrsOrWarning {
+    Attributes(TestcaseAttrs),
+    Warning(String),
 }
 
-// from https://gist.github.com/scott-codecov/311c174ecc7de87f7d7c50371c6ef927#file-cobertura-rs-L18-L31
-fn get_relevant_attrs(attributes: Attributes) -> PyResult<RelevantAttrs> {
-    let mut rel_attrs = RelevantAttrs::default();
+// originally from https://gist.github.com/scott-codecov/311c174ecc7de87f7d7c50371c6ef927#file-cobertura-rs-L18-L31
+fn parse_testcase_attrs(attributes: Attributes) -> Result<AttrsOrWarning> {
```

Comment:
instead of having an enum for this, have we considered creating a specific `Err` type for "we're missing some info" and matching the error returned from the `Result` to that error type so we know some info is missing?

The caller of `parse_testcase_attrs` can infer if parsing attributes failed with a "recuperable" error. Would let you keep the nice interface of `RelevantAttrs` and maybe simplify the code a little bit

Reply:
I agree, that would make things here a bit cleaner.